plugins/source/hackernews/resources/services/items/items_fetch.go

```diff
@@ -0,0 +1,114 @@
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/cloudquery/cloudquery/plugins/source/hackernews/client"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/hermanschaaf/hackernews"
+	"golang.org/x/sync/errgroup"
+)
+
+// fetchItems gets called by the CloudQuery SDK to fetch the data for the `hackernews_items` table. It is passed a context and a
+// schema.ClientMeta interface which is used to access the plugin configuration and the state backend. The
+// resource argument is not used here, as it refers to the parent table, which is not applicable in this case.
+// The res channel is used to send the fetched data back to the CloudQuery SDK.
+//
+// Because this is an incremental table, we must first load the cursor from the state backend, and then update
+// it whenever we've successfully fetched a new item. The cursor is a string, and in this case it is the ID of
+// the last item we've fetched.
+//
+// Incremental tables should guarantee at-least-once delivery. The Hacker News API gives us the current maximum item ID,
+// and we need to fetch all items up to that ID. Once we've fetched all items up to a certain ID, we can safely
+// update the cursor. If we crash before updating the cursor, we will fetch the same items again on the next run.
+//
+// This resolver is different from most in that it needs to concurrently fetch items using its own goroutines. Usually,
+// a resolver with pagination will not have this ability and will simply iterate over all pages.
+func fetchItems(ctx context.Context, meta schema.ClientMeta, _ *schema.Resource, res chan<- any) error {
+	c := meta.(*client.Client)
+	tableName := Items().Name
+	value, err := c.Backend.Get(ctx, tableName, c.ID())
+	if err != nil {
+		return fmt.Errorf("failed to retrieve state from backend: %w", err)
+	}
+
+	// read the cursor from the state, or default to 0 if it's not set
+	cursor := 0
+	if value == "" {
+		c.Logger().Info().
+			Str("table", tableName).Str("client_id", c.ID()).Msgf("No previous cursor found")
+	} else {
+		cursor, err = strconv.Atoi(value)
+		if err != nil {
+			return fmt.Errorf("failed to convert cursor to int: %w", err)
+		}
+		c.Logger().Info().
+			Str("table", tableName).Str("client_id", c.ID()).Msg("Found previous cursor with value " + strconv.Itoa(cursor))
+	}
+
+	// find the max item ID from the Hacker News API
+	maxID, err := c.HackerNews.MaxItemID(ctx)
+	if err != nil {
+		return err
+	}
+	c.Logger().Info().Msg("Found max ID, reading up to " + strconv.Itoa(maxID))
+
+	// Fetch items in batches of (max) 1000.
+	// This is not necessarily the most efficient way of doing it, but this code
+	// is meant to be for instructional purposes as an example of updating cursors,
+	// so we're keeping the logic relatively simple.
+	// The important thing is that the state backend does not ensure that the cursor
+	// is strictly increasing--it is the responsibility of the resolver to ensure this.
+	for cursor < maxID {
+		endID := cursor + 1000
```

Comment:
```suggestion
		endID := cursor + c.Spec.ItemConcurrency
```

If we do this we won't have to `g.SetLimit(c.Spec.ItemConcurrency)`. Also we keep the batches as small as possible so lower chances of a batch to fail

Reply:
I tried this, but it has a pretty big impact on performance, even after playing with different values for `ItemConcurrency` (on my machine: ~500 items/s falls to ~300 items/s). I think let's leave it at 1000 for now